File: LUA/flynn.lua
local websocket = require('websocket')
local json = require('json')
local fs = require('fs')
local readline = require('readline')
local http = require('http')

local cl = { gr = '\x1b[32m', gb = '\x1b[4m', br = '\x1b[34m', st = '\x1b[9m', yl = '\x1b[33m', rt = '\x1b[0m' }

local socket = nil
local pingInterval
local countdownInterval
local potentialPoints = 0
local countdown = "Calculating..."
local pointsTotal = 0
local pointsToday = 0
local reconnectAttempts = 0
local maxReconnectAttempts = 5
local maxReconnectInterval = 5 * 60 * 1000 -- 5 minutes in milliseconds
local CoderMarkPrinted = false

function CoderMark()
    if not CoderMarkPrinted then
        print([[
╭━━━╮╱╱╱╱╱╱╱╱╱╱╱╱╱╭━━━┳╮
┃╭━━╯╱╱╱╱╱╱╱╱╱╱╱╱╱┃╭━━┫┃]] .. cl.gr .. [[
┃╰━━┳╮╭┳━┳━━┳━━┳━╮┃╰━━┫┃╭╮╱╭┳━╮╭━╮
┃╭━━┫┃┃┃╭┫╭╮┃╭╮┃╭╮┫╭━━┫┃┃┃╱┃┃╭╮┫╭╮╮]] .. cl.br .. [[
┃┃╱╱┃╰╯┃┃┃╰╯┃╰╯┃┃┃┃┃╱╱┃╰┫╰━╯┃┃┃┃┃┃┃
╰╯╱╱╰━━┻╯╰━╮┣━━┻╯╰┻╯╱╱╰━┻━╮╭┻╯╰┻╯╰╯]] .. cl.rt .. [[
╱╱╱╱╱╱╱╱╱╱╱┃┃╱╱╱╱╱╱╱╱╱╱╱╭━╯┃
╱╱╱╱╱╱╱╱╱╱╱╰╯╱╱╱╱╱╱╱╱╱╱╱╰━━╯
\n]] .. cl.gb .. [[Teneo Node Cli ]] .. cl.gr .. [[v1.1.0 ]] .. cl.rt .. cl.gb .. [[dev_build]] .. cl.rt .. [[
        ]])
        CoderMarkPrinted = true
    end
end

local function readFileAsync(filename)
    local file, err = fs.readFileSync(filename, 'utf8')
    if err then return {} end
    return json.decode(file)
end

local function writeFileAsync(filename, data)
    fs.writeFileSync(filename, json.encode(data))
end

local rl = readline.createInterface({
    input = io.stdin,
    output = io.stdout
})

local function getLocalStorage()
    local success, data = pcall(readFileAsync, 'localStorage.json')
    return success and data or {}
end

local function setLocalStorage(data)
    local currentData = getLocalStorage()
    local newData = {}
    for k, v in pairs(currentData) do newData[k] = v end
    for k, v in pairs(data) do newData[k] = v end
    writeFileAsync('localStorage.json', newData)
end

local function getUserIdFromFile()
    local success, data = pcall(readFileAsync, 'UserId.json')
    return success and data.userId or nil
end

local function setUserIdToFile(userId)
    writeFileAsync('UserId.json', { userId = userId })
end

local function getAccountData()
    local success, data = pcall(readFileAsync, 'DataAccount.json')
    return success and data or {}
end

local function setAccountData(email, password, access_token, refresh_token, personalCode)
    local accountData = {
        email = email,
        password = password,
        access_token = access_token,
        refresh_token = refresh_token,
        personalCode = personalCode
    }
    writeFileAsync('DataAccount.json', accountData)
end

local function getReconnectDelay(attempt)
    local baseDelay = 5000 -- 5 seconds
    local additionalDelay = attempt * 5000 -- Additional 5 seconds for each attempt
    return math.min(baseDelay + additionalDelay, maxReconnectInterval)
end

local function connectWebSocket(userId)
    if socket then return end
    local version = "v0.2"
    local url = "wss://secure.ws.teneo.pro"
    local wsUrl = url .. "/websocket?userId=" .. http.urlEncode(userId) .. "&version=" .. http.urlEncode(version)
    socket = websocket.client()

    socket:on("connect", function()
        reconnectAttempts = 0
        local connectionTime = os.date("!%Y-%m-%dT%H:%M:%SZ")
        setLocalStorage({ lastUpdated = connectionTime })
        print("WebSocket connected at", connectionTime)
        startPinging()
        startCountdownAndPoints()
    end)

    socket:on("message", function(event)
        local data = json.decode(event)
        print("Received message from WebSocket:", data)
        if data.pointsTotal and data.pointsToday then
            local lastUpdated = os.date("!%Y-%m-%dT%H:%M:%SZ")
            setLocalStorage({
                lastUpdated = lastUpdated,
                pointsTotal = data.pointsTotal,
                pointsToday = data.pointsToday,
            })
            pointsTotal = data.pointsTotal
            pointsToday = data.pointsToday
        end
    end)

    socket:on("close", function(event)
        socket = nil
        print("WebSocket disconnected")
        stopPinging()
        if not event.wasClean then
            reconnectAttempts = reconnectAttempts + 1
            local delay = getReconnectDelay(reconnectAttempts)
            if delay < maxReconnectInterval then
                timer.sleep(delay)
                reconnectWebSocket()
            else
                print("Max reconnect interval reached. Giving up.")
            end
        end
    end)

    socket:on("error", function(error)
        print("WebSocket error:", error)
    end)

    socket:connect(wsUrl)
end

local function disconnectWebSocket()
    if socket then
        socket:close()
        socket = nil
        stopPinging()
    end
end

local function startPinging()
    stopPinging()
    pingInterval = timer.setInterval(function()
        if socket and socket.readyState == websocket.OPEN then
            socket:send(json.encode({ type = "PING" }))
            setLocalStorage({ lastPingDate = os.date("!%Y-%m-%dT%H:%M:%SZ") })
        end
    end, 10000)
end

local function stopPinging()
    if pingInterval then
        timer.clearInterval(pingInterval)
        pingInterval = nil
    end
end

local function startCountdownAndPoints()
    timer.clearInterval(countdownInterval)
    updateCountdownAndPoints()
    countdownInterval = timer.setInterval(updateCountdownAndPoints, 1000)
end

local function updateCountdownAndPoints()
    local localStorage = getLocalStorage()
    local lastUpdated = localStorage.lastUpdated
    if lastUpdated then
        local nextHeartbeat = os.date("!*t", os.time(lastUpdated) + 15 * 60)
        local now = os.date("!*t")
        local diff = os.time(nextHeartbeat) - os.time(now)

        if diff > 0 then
            local minutes = math.floor(diff / 60)
            local seconds = diff % 60
            countdown = string.format("%dm %ds", minutes, seconds)

            local maxPoints = 25
            local timeElapsed = os.time(now) - os.time(lastUpdated)
            local timeElapsedMinutes = timeElapsed / 60
            local newPoints = math.min(maxPoints, (timeElapsedMinutes / 15) * maxPoints)
            newPoints = tonumber(string.format("%.2f", newPoints))

            if math.random() < 0.1 then
                local bonus = math.random() * 2
                newPoints = math.min(maxPoints, newPoints + bonus)
                newPoints = tonumber(string.format("%.2f", newPoints))
            end

            potentialPoints = newPoints
        else
            countdown = "Calculating..."
            potentialPoints = 25
        end
    else
        countdown = "Calculating..."
        potentialPoints = 0
    end
    setLocalStorage({ potentialPoints = potentialPoints, countdown = countdown })
end

local function getUserId()
    local loginUrl = "https://node-community-api.teneo.pro/auth/v1/token?grant_type=password"
    local authorization = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag"
    local apikey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag"

    local accountData = getAccountData()
    local email = accountData.email or rl:question("\nEmail: " .. cl.gr)
    local password = accountData.password or rl:question(cl.rt .. "Password: " .. cl.st .. cl.br .. "", function(input) return input end)

    local response, err = http.post(loginUrl, json.encode({ email = email, password = password }), {
        headers = {
            Authorization = authorization,
            apikey = apikey
        }
    })

    if response then
        local data = json.decode(response.body)
        local access_token = data.access_token
        local refresh_token = data.refresh_token
        print('Access_Token:', access_token)
        print('Refresh_Token:', refresh_token)

        local AuthUserUrl = "https://node-community-api.teneo.pro/auth/v1/user"
        local AuthResponse = http.get(AuthUserUrl, {
            headers = {
                Authorization = "Bearer " .. access_token,
                apikey = apikey
            }
        })

        local userId = json.decode(AuthResponse.body).id
        print('User ID:', userId)

        local profileUrl = "https://node-community-api.teneo.pro/rest/v1/profiles?select=personal_code&id=eq." .. userId
        local profileResponse = http.get(profileUrl, {
            headers = {
                Authorization = "Bearer " .. access_token,
                apikey = apikey
            }
        })
        local personalCode = json.decode(profileResponse.body)[1].personal_code
        print("Personal Code:" .. cl.rt, personalCode)
        setUserIdToFile(userId)
        setAccountData(email, password, access_token, refresh_token, personalCode)
        startCountdownAndPoints()
        connectWebSocket(userId)
        io.write(cl.gr .. "Data has been saved in the DataAccount.json file...\n" .. cl.rt)
        CoderMark()
    else
        print('Error:', err)
    end
end

local function reconnectWebSocket()
    local userId = getUserIdFromFile()
    if userId then
        connectWebSocket(userId)
    end
end

local function autoLogin()
    local accountData = getAccountData()
    if accountData.email and accountData.password then
        getUserId()
        print(cl.yl .. "\nAutomatic Login has been Successfully Executed..\n" .. cl.rt)
    end
end

local function main()
    local localStorageData = getLocalStorage()
    local userId = getUserIdFromFile()

    if not userId then
        rl:question("\nUser ID not found. Would you like to:\n" .. cl.gr .. "\n1. Login to your account\n2. Enter User ID manually\n" .. cl.rt .. "\nChoose an option: ", function(option)
            if option == '1' then
                getUserId()
            elseif option == '2' then
                rl:question('Please enter your user ID: ', function(inputUserId)
                    userId = inputUserId
                    setUserIdToFile(userId)
                    startCountdownAndPoints()
                    connectWebSocket(userId)
                    rl:close()
                end)
            else
                print('Invalid option. Exiting...')
                os.exit(0)
            end
        end)
    else
        rl:question("\nMenu:\n" .. cl.gr .. "\n1. Logout\n2. Start Running Node\n" .. cl.rt .. "\nChoose an option: ", function(option)
            if option == '1' then
                fs.unlink('UserId.json', function(err) if err then error(err) end end)
                fs.unlink('localStorage.json', function(err) if err then error(err) end end)
                fs.unlink('DataAccount.json', function(err) if err then error(err) end end)
                print(cl.yl .. "\nLogged out successfully.")
                os.exit(0)
            elseif option == '2' then
                os.execute("clear")
                CoderMark()
                print(cl.gr .. "Initiates a connection to the node...\n" .. cl.rt)
                startCountdownAndPoints()
                connectWebSocket(userId)
            else
                print(cl.yl .. "Invalid option. Exiting...")
                os.exit(0)
            end
        end)
    end
end

main()
timer.setInterval(autoLogin, 1800000)

File: NODEJS/flynn.js
const WebSocket = require('ws');
const { promisify } = require('util');
const fs = require('fs');
const readline = require('readline');
const axios = require('axios');

const cl = { gr: '\x1b[32m', gb: '\x1b[4m', br: '\x1b[34m', st: '\x1b[9m', yl: '\x1b[33m', rt: '\x1b[0m' };

let socket = null;
let pingInterval;
let countdownInterval;
let potentialPoints = 0;
let countdown = "Calculating...";
let pointsTotal = 0;
let pointsToday = 0;
let reconnectAttempts = 0;
const maxReconnectAttempts = 5;
const maxReconnectInterval = 5 * 60 * 1000; // 5 minutes in milliseconds
let CoderMarkPrinted = false;

function CoderMark() {
    if (!CoderMarkPrinted) {
        console.log(`
╭━━━╮╱╱╱╱╱╱╱╱╱╱╱╱╱╭━━━┳╮
┃╭━━╯╱╱╱╱╱╱╱╱╱╱╱╱╱┃╭━━┫┃${cl.gr}
┃╰━━┳╮╭┳━┳━━┳━━┳━╮┃╰━━┫┃╭╮╱╭┳━╮╭━╮
┃╭━━┫┃┃┃╭┫╭╮┃╭╮┃╭╮┫╭━━┫┃┃┃╱┃┃╭╮┫╭╮╮${cl.br}
┃┃╱╱┃╰╯┃┃┃╰╯┃╰╯┃┃┃┃┃╱╱┃╰┫╰━╯┃┃┃┃┃┃┃
╰╯╱╱╰━━┻╯╰━╮┣━━┻╯╰┻╯╱╱╰━┻━╮╭┻╯╰┻╯╰╯${cl.rt}
╱╱╱╱╱╱╱╱╱╱╱┃┃╱╱╱╱╱╱╱╱╱╱╱╭━╯┃
╱╱╱╱╱╱╱╱╱╱╱╰╯╱╱╱╱╱╱╱╱╱╱╱╰━━╯
\n${cl.gb}Teneo Node Cli ${cl.gr}v1.1.0 ${cl.rt}${cl.gb}${cl.br}dev_build${cl.rt}
        `);
        CoderMarkPrinted = true;
    }
}

const readFileAsync = promisify(fs.readFile);
const writeFileAsync = promisify(fs.writeFile);

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

async function getLocalStorage() {
    try {
        const data = await readFileAsync('localStorage.json', 'utf8');
        return JSON.parse(data);
    } catch (error) {
        return {};
    }
}

async function setLocalStorage(data) {
    const currentData = await getLocalStorage();
    const newData = {
        ...currentData,
        ...data
    };
    await writeFileAsync('localStorage.json', JSON.stringify(newData));
}

async function getUserIdFromFile() {
    try {
        const data = await readFileAsync('UserId.json', 'utf8');
        return JSON.parse(data).userId;
    } catch (error) {
        return null;
    }
}

async function setUserIdToFile(userId) {
    await writeFileAsync('UserId.json', JSON.stringify({
        userId
    }));
}

async function getAccountData() {
    try {
        const data = await readFileAsync('DataAccount.json', 'utf8');
        return JSON.parse(data);
    } catch (error) {
        return {};
    }
}

async function setAccountData(email, password, access_token, refresh_token, personalCode) {
    const accountData = {
        email,
        password,
        access_token,
        refresh_token,
        personalCode
    };
    await writeFileAsync('DataAccount.json', JSON.stringify(accountData));
}

function getReconnectDelay(attempt) {
    const baseDelay = 5000; // 5 seconds
    const additionalDelay = attempt * 5000; // Additional 5 seconds for each attempt
    return Math.min(baseDelay + additionalDelay, maxReconnectInterval);
}

async function connectWebSocket(userId) {
    if (socket) return;
    const version = "v0.2";
    const url = "wss://secure.ws.teneo.pro";
    const wsUrl = `${url}/websocket?userId=${encodeURIComponent(userId)}&version=${encodeURIComponent(version)}`;
    socket = new WebSocket(wsUrl);

    socket.onopen = async () => {
        reconnectAttempts = 0;
        const connectionTime = new Date().toISOString();
        await setLocalStorage({
            lastUpdated: connectionTime
        });
        console.log("WebSocket connected at", connectionTime);
        startPinging();
        startCountdownAndPoints();
    };

    socket.onmessage = async (event) => {
        const data = JSON.parse(event.data);
        console.log("Received message from WebSocket:", data);
        if (data.pointsTotal !== undefined && data.pointsToday !== undefined) {
            const lastUpdated = new Date().toISOString();
            await setLocalStorage({
                lastUpdated: lastUpdated,
                pointsTotal: data.pointsTotal,
                pointsToday: data.pointsToday,
            });
            pointsTotal = data.pointsTotal;
            pointsToday = data.pointsToday;
        }
    };

    socket.onclose = (event) => {
        socket = null;
        console.log("WebSocket disconnected");
        stopPinging();
        if (!event.wasClean) {
            reconnectAttempts++;
            const delay = getReconnectDelay(reconnectAttempts);
            if (delay < maxReconnectInterval) {
                setTimeout(() => {
                    reconnectWebSocket();
                }, delay);
            } else {
                console.log("Max reconnect interval reached. Giving up.");
            }
        }
    };

    socket.onerror = (error) => {
        console.error("WebSocket error:", error);
    };
}

function disconnectWebSocket() {
    if (socket) {
        socket.close();
        socket = null;
        stopPinging();
    }
}

function startPinging() {
    stopPinging();
    pingInterval = setInterval(async () => {
        if (socket && socket.readyState === WebSocket.OPEN) {
            socket.send(JSON.stringify({
                type: "PING"
            }));
            await setLocalStorage({
                lastPingDate: new Date().toISOString()
            });
        }
    }, 10000);
}

function stopPinging() {
    if (pingInterval) {
        clearInterval(pingInterval);
        pingInterval = null;
    }
}

process.on('SIGINT', () => {
    console.log('Received SIGINT. Stopping pinging...');
    stopPinging();
    disconnectWebSocket();
    process.exit(0);
});

function startCountdownAndPoints() {
    clearInterval(countdownInterval);
    updateCountdownAndPoints();
    countdownInterval = setInterval(updateCountdownAndPoints, 1000);
}

async function updateCountdownAndPoints() {
    const {
        lastUpdated
    } = await getLocalStorage();
    if (lastUpdated) {
        const nextHeartbeat = new Date(lastUpdated);
        nextHeartbeat.setMinutes(nextHeartbeat.getMinutes() + 15);
        const now = new Date();
        const diff = nextHeartbeat.getTime() - now.getTime();

        if (diff > 0) {
            const minutes = Math.floor(diff / 60000);
            const seconds = Math.floor((diff % 60000) / 1000);
            countdown = `${minutes}m ${seconds}s`;

            const maxPoints = 25;
            const timeElapsed = now.getTime() - new Date(lastUpdated).getTime();
            const timeElapsedMinutes = timeElapsed / (60 * 1000);
            let newPoints = Math.min(maxPoints, (timeElapsedMinutes / 15) * maxPoints);
            newPoints = parseFloat(newPoints.toFixed(2));

            if (Math.random() < 0.1) {
                const bonus = Math.random() * 2;
                newPoints = Math.min(maxPoints, newPoints + bonus);
                newPoints = parseFloat(newPoints.toFixed(2));
            }

            potentialPoints = newPoints;
        } else {
            countdown = "Calculating...";
            potentialPoints = 25;
        }
    } else {
        countdown = "Calculating...";
        potentialPoints = 0;
    }
    await setLocalStorage({
        potentialPoints,
        countdown
    });
}

async function getUserId() {
    const loginUrl = "https://node-community-api.teneo.pro/auth/v1/token?grant_type=password";
    const authorization = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag";
    const apikey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag";

    const accountData = await getAccountData();
    const email = accountData.email || await new Promise(resolve => rl.question(`\nEmail: ` + cl.gr, resolve));
    const password = accountData.password || await new Promise(resolve => rl.question(cl.rt + `Password: ` + cl.st + cl.br + ``, resolve));

    try {
        const response = await axios.post(loginUrl, {
            email: email,
            password: password
        }, {
            headers: {
                'Authorization': authorization,
                'apikey': apikey
            }
        });

        const access_token = response.data.access_token;
        const refresh_token = response.data.refresh_token;
        console.log('Access_Token:', access_token);
        console.log('Refresh_Token:', refresh_token);

        const AuthUserUrl = "https://node-community-api.teneo.pro/auth/v1/user";
        const AuthResponse = await axios.get(AuthUserUrl, {
            headers: {
                'Authorization': `Bearer ${access_token}`,
                'apikey': apikey
            }
        });

        const userId = AuthResponse.data.id;
        console.log('User ID:', userId);

        const profileUrl = `https://node-community-api.teneo.pro/rest/v1/profiles?select=personal_code&id=eq.${userId}`;
        const profileResponse = await axios.get(profileUrl, {
            headers: {
                'Authorization': `Bearer ${access_token}`,
                'apikey': apikey
            }
        });
        const personalCode = profileResponse.data[0]?.personal_code;
        console.log(`Personal Code:` + cl.rt, personalCode);
        await setUserIdToFile(userId);
        await setAccountData(email, password, access_token, refresh_token, personalCode);
        await startCountdownAndPoints();
        await connectWebSocket(userId);
        console.clear();
        console.log(cl.gr + `Data has been saved in the DataAccount.json file...\n` + cl.rt);
        CoderMark();
    } catch (error) {
        console.error('Error:', error.response ? error.response.data : error.message);
    } finally {
        rl.close();
    }
}

async function reconnectWebSocket() {
    const userId = await getUserIdFromFile();
    if (userId) {
        await connectWebSocket(userId);
    }
}

async function autoLogin() {
    const accountData = await getAccountData();
    if (accountData.email && accountData.password) {
        await getUserId();
        console.log(cl.yl + `\nAutomatic Login has been Successfully Executed..\n` + cl.rt);
    }
}

async function main() {
    const localStorageData = await getLocalStorage();
    let userId = await getUserIdFromFile();

    if (!userId) {
        rl.question(`\nUser ID not found. Would you like to:\n` + cl.gr + `\n1. Login to your account\n2. Enter User ID manually\n` + cl.rt + `\nChoose an option: `, async (option) => {
            switch (option) {
                case '1':
                    await getUserId();
                    break;
                case '2':
                    rl.question('Please enter your user ID: ', async (inputUserId) => {
                        userId = inputUserId;
                        await setUserIdToFile(userId);
                        await startCountdownAndPoints();
                        await connectWebSocket(userId);
                        rl.close();
                    });
                    break;
                default:
                    console.log('Invalid option. Exiting...');
                    process.exit(0);
            }
        });
    } else {
        rl.question(`\nMenu:\n` + cl.gr + `\n1. Logout\n2. Start Running Node\n` + cl.rt + `\nChoose an option: `, async (option) => {
            switch (option) {
                case '1':
                    fs.unlink('UserId.json', (err) => {
                        if (err) throw err;
                    });
                    fs.unlink('localStorage.json', (err) => {
                        if (err) throw err;
                    });
                    fs.unlink('DataAccount.json', (err) => {
                        if (err) throw err;
                    });
                    console.log(cl.yl + `\nLogged out successfully.`);
                    process.exit(0);
                    break;
                case '2':
                    console.clear();
                    CoderMark();
                    console.log(cl.gr + `Initiates a connection to the node...\n` + cl.rt);
                    await startCountdownAndPoints();
                    await connectWebSocket(userId);
                    break;
                default:
                    console.log(cl.yl + `Invalid option. Exiting...`);
                    process.exit(0);
            }
        });
    }
}

main();
setInterval(autoLogin, 1800000);

